lang/LangPrimSource/SC_AudioDevicePrim.cpp

```diff
@@ -125,6 +131,56 @@ int listDevices(struct VMGlobals* g, int type) {
     return 1;
 }
 
+#elif defined(SC_AUDIO_API_PORTAUDIO)
+int listDevices(struct VMGlobals* g, int type) {
+    int numDevices;
+    PaError err = Pa_Initialize();
+    if (err != paNoError)
+        return 0;
+
+    numDevices = Pa_GetDeviceCount();
+    if (numDevices < 0)
+        return 0;
+
+    std::vector<const PaDeviceInfo*> deviceInfos;
+    for (PaDeviceIndex i = 0; i < numDevices; i++) {
+        auto* pdi = Pa_GetDeviceInfo(i);
+        if (type < BOTH) {
+            if (type < IN) {
+                if (pdi->maxOutputChannels > 0)
+                    deviceInfos.push_back(pdi);
+            } else {
+                if (pdi->maxInputChannels > 0)
+                    deviceInfos.push_back(pdi);
+            }
+        } else
+            deviceInfos.push_back(pdi);
+    }
+
+    int numSelectedDevices = deviceInfos.size();
+
+    PyrSlot* a = g->sp - 2;
+    PyrObject* devArray = newPyrArray(g->gc, numSelectedDevices * sizeof(PyrObject), 0, true);
+    SetObject(a, devArray); // this is okay here as we don't use the receiver below
+
+    for (int i = 0; i < numSelectedDevices; i++) {
+        PyrString* string = newPyrString(g->gc, GetPaDeviceName(deviceInfos[i]).c_str(), 0, true);
+        SetObject(devArray->slots + i, string);
+        devArray->size++;
+        g->gc->GCWriteNew(devArray, (PyrObject*)string); // we know array is white so we can use GCWriteNew
```

Comment:
generally try to avoid c-style casts when a less aggressive cast will work -- c-style casts can do lots of conversion including changing const-ness (https://en.cppreference.com/w/cpp/language/explicit_cast). here, you should use `static_cast<PyrObject*>(string)`

Reply:
I can't say that I understand all these particular objects (`PyrString`, `PyrObject`) and this was copied from the coreaudio code. However, when I tried to use your suggestion, I got 
```
error: static_cast
      from 'PyrString *' to 'PyrObject *', which are not related by inheritance, is not allowed
```

Reply:
ah, my mistake, i thought PyrString was a derived class of PyrObject, but it's not. you'll have to use `reinterpret_cast`.